units/migrate_volume_job.go

```diff
@@ -0,0 +1,286 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/cloud"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const (
+	volumeMigrationJobName = "volume-migration"
+)
+
+func init() {
+	registry.AddJobType(volumeMigrationJobName,
+		func() amboy.Job { return makeVolumeMigrationJob() })
+}
+
+type volumeMigrationJob struct {
+	job.Base      `bson:"job_base" json:"job_base" yaml:"job_base"`
+	VolumeID      string             `bson:"volume_id" yaml:"volume_id"`
+	ModifyOptions cloud.SpawnOptions `bson:"modify_options" json:"modify_options" yaml:"modify_options"`
+
+	VolumeDetached     bool   `bson:"volume_detached"`
+	InitialHostID      string `bson:"initial_host_id"`
+	SpawnhostStopJobID string `bson:"spawnhost_stop_job_id"`
+	NewHostID          string `bson:"new_host_id"`
+
+	env         evergreen.Environment
+	volume      *host.Volume
+	initialHost *host.Host
+}
+
+func makeVolumeMigrationJob() *volumeMigrationJob {
+	j := &volumeMigrationJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    volumeMigrationJobName,
+				Version: 0,
+			},
+		},
+	}
+	return j
+}
+
+func NewVolumeMigrationJob(env evergreen.Environment, volumeID string, modifyOptions cloud.SpawnOptions, ts string) amboy.Job {
+	j := makeVolumeMigrationJob()
+	j.env = env
+	j.SetID(fmt.Sprintf("%s.%s.%s", volumeMigrationJobName, volumeID, ts))
+	j.SetScopes([]string{fmt.Sprintf("%s.%s", volumeMigrationJobName, volumeID)})
+	j.SetEnqueueAllScopes(true)
+	j.VolumeID = volumeID
+	j.ModifyOptions = modifyOptions
+	j.UpdateRetryInfo(amboy.JobRetryOptions{
+		Retryable:   utility.TruePtr(),
+		MaxAttempts: utility.ToIntPtr(30),
+		WaitUntil:   utility.ToTimeDurationPtr(20 * time.Second),
+	})
+
+	return j
+}
+
+func (j *volumeMigrationJob) Run(ctx context.Context) {
+	defer j.finishJob(ctx)
+
+	if err := j.populateIfUnset(); err != nil {
+		j.AddRetryableError(err)
+		return
+	}
+
+	if err := j.volume.SetMigrating(true); err != nil {
```

Comment:
Its main purpose is to be used by the UI! Since most of the operations are idempotent and we're checking the host/volume documents throughout the job there is nothing that truly needs to be locked via this field.